toolchain/lex/lex.cpp

```diff
@@ -0,0 +1,1309 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/lex/lex.h"
+
+#include <array>
+
+#include "common/check.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/StringSwitch.h"
+#include "llvm/Support/Compiler.h"
+#include "toolchain/base/value_store.h"
+#include "toolchain/lex/character_set.h"
+#include "toolchain/lex/helpers.h"
+#include "toolchain/lex/numeric_literal.h"
+#include "toolchain/lex/string_literal.h"
+#include "toolchain/lex/tokenized_buffer.h"
+
+#if __ARM_NEON
+#include <arm_neon.h>
+#define CARBON_USE_SIMD 1
+#elif __x86_64__
+#include <x86intrin.h>
+#define CARBON_USE_SIMD 1
+#else
+#define CARBON_USE_SIMD 0
+#endif
+
+namespace Carbon::Lex {
+
+// Implementation of the lexer logic itself.
+//
+// The design is that lexing can loop over the source buffer, consuming it into
+// tokens by calling into this API. This class handles the state and breaks down
+// the different lexing steps that may be used. It directly updates the provided
+// tokenized buffer with the lexed tokens.
+//
+// Because of the internal linkage attribute, this must be in the cpp file where
+// it's used.
+// TODO: Why is this needed?
```

Comment:
Suggested comment to explain the use of internal linkage:

```suggestion
// We'd typically put this in an anonymous namespace, but it is `friend`-ed by
// the `TokenizedBuffer`. One of the important benefits of being in an anonymous
// namespace is having internal linkage. That allows the optimizer to much more
// aggressively inline away functions that are called in only one place. We keep
// that benefit for now by using the `internal_linkage` attribute.
//
// TODO: Investigate ways to refactor the code that allow moving this into an
// anonymous namespace without overly exposing implementation details of the
// `TokenizedBuffer` or undermining the performance constraints of the lexer.
```